segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ReadOperation.java

```diff
@@ -141,23 +143,56 @@ private void logEnd() {
-                                                    currentOffset.addAndGet(bytesRead);
-                                                    currentBufferOffset.addAndGet(bytesRead);
-                                                    totalBytesRead.addAndGet(bytesRead);
-                                                }, chunkedSegmentStorage.getExecutor()),
-                                        chunkedSegmentStorage.getExecutor());
+                        return CompletableFuture.runAsync(() -> {
+                            // Create parallel requests to read each chunk.
+                            chunkReadFutures.add(readChunk(chunkToReadFrom.getName(),
+                                    currentOffset.get() - startOffsetForCurrentChunk.get(),
+                                    bytesToRead,
+                                    currentBufferOffset.get()));
+                            log.trace("{} read - reading chunk - op={}, segment={}, chunk={} offset={} length={} bufferOffset={}",
+                                    chunkedSegmentStorage.getLogPrefix(),
+                                    System.identityHashCode(this), handle.getSegmentName(),
+                                    chunkToReadFrom.getName(),
+                                    currentOffset.get() - startOffsetForCurrentChunk.get(),
+                                    bytesToRead,
+                                    currentBufferOffset.get());
+                            bytesRemaining.addAndGet(-bytesToRead);
+                            currentOffset.addAndGet(bytesToRead);
+                            currentBufferOffset.addAndGet(bytesToRead);
+                            totalBytesRead.addAndGet(bytesToRead);
+                        }, chunkedSegmentStorage.getExecutor());
                     }
                     return CompletableFuture.completedFuture(null);
-                }, chunkedSegmentStorage.getExecutor());
+                }, chunkedSegmentStorage.getExecutor())
+                .thenCompose(v -> Futures.allOf(chunkReadFutures));
+    }
+
+    private CompletableFuture<Void> readChunk(String chunkName,
+                         long fromOffset,
+                         int bytesToRead,
+                         int bufferOffset) {
+        val chunkBytesRemaining = new AtomicInteger(bytesToRead);
+        val chunkFromOffset = new AtomicLong(fromOffset);
+        val chunkBufferOffset = new AtomicInteger(bufferOffset);
+        // Note that technically it is possible that read actually request reads less than requested bytes, requiring additional reads on the same chunk.
+        // Hence the for loop below.
+        return chunkedSegmentStorage.getChunkStorage().openRead(chunkName)
+                .thenComposeAsync(chunkHandle ->
+                    Futures.loop(
+                            () -> chunkBytesRemaining.get() > 0 && null != currentChunkName,
```

Comment:
where is `currentChunkName` defined and/or updated?

Reply:
Ok, I see it's a class-level variable. Maybe let's add some comment explaining this condition and how the loop will eventually terminate.

Reply:
actually `null != currentChunkName` is not needed here is that condition is already checked before calling this method.